### PR TITLE
update unify cycles

### PR DESCRIPTION
**Pull request by romanarust**
As noted in https://github.com/compas-dev/compas/issues/1411 there is a problem with face_adjacency on meshes with arbitrary-sized faces.  I've now exposed the parameters `radius` and `nmax` from `compas.topology._face_adjacency` to `compas.topology.face_adjacency` and further propagated them to `unify_cycles` and `Mesh.unify_cycles`.

### What type of change is this?

- [x] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas/datastructures/mesh/mesh.py [hunk 1]
```diff
@@ -2932,9 +2932,18 @@ def quads_to_triangles(self, check_angles=False):
                     del self.facedata[face]
 
     # only reason this is here and not on the halfedge is because of the spatial tree
-    def unify_cycles(self, root=None):
+    def unify_cycles(self, root=None, nmax=None, radius=None):
         """Unify the cycles of the mesh.
 
+        Parameters
+        ----------
+        root : str, optional
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/mesh.py]: type of root is `int`

File: src/compas/topology/orientation.py [hunk 2]
```diff
@@ -166,6 +172,10 @@ def unify_cycles(vertices, faces, root=None):
         The faces of the mesh defined as lists of vertex indices.
     root : str, optional
```
Comment (tomvanmele) [inline src/compas/topology/orientation.py]: same here

File: src/compas/topology/orientation.py [hunk 3]
```diff
@@ -117,10 +121,12 @@ def face_adjacency(points, faces):
...
-    f = len(faces)
-
-    if f > 100:
-        return _face_adjacency(points, faces)
+    if nmax is not None:
+        if radius is not None:
+            return _face_adjacency(points, faces, nmax=nmax, radius=radius)
+        return _face_adjacency(points, faces, nmax=nmax)
+    if radius is not None:
+        return _face_adjacency(points, faces, radius=radius)
```
Comment (tomvanmele) [inline src/compas/topology/orientation.py]: perhaps it would make sense to refactor `_face_adjacency` such that you can write here
```python
if nmax or radius:
    return _face_adjacency(points, faces, nmax=nmax, radius=radius)
```

**Comment by tomvanmele** [comment]
btw, since the switch is no longer based on the number of faces, it should be easier to test using generated boundary cases...

**Comment by romanarust** [comment]
> btw, since the switch is no longer based on the number of faces, it should be easier to test using generated boundary cases...

correct. do you have something in mind or should I leave the tests for now?

**Comment by tomvanmele** [comment]
was thinking that we could use the meshgrid with awkward proportions in x and y to generate inputs where radius and nmax provide very different results in different directions...

**Review by tomvanmele — DISMISSED** [review]
LGTM

File: tests/compas/topology/test_unify_cycles.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def test_unify_cycles():
+    if compas.IPY:
+        return
+    test_data = compas.json_load(os.path.join(HERE, "..", "fixtures", "topology", "vertices_faces.json"))
+    vertices = test_data["vertices"]
+    faces = test_data["faces"]
+    unify_cycles(vertices, faces)
+    unify_cycles(vertices, faces, nmax=29, max_distance=22.4)  # anything below won't work
```
Comment (tomvanmele) [inline tests/compas/topology/test_unify_cycles.py]: is it clear why that is the case?
or it is just something you noticed?

File: tests/compas/topology/test_unify_cycles.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def test_face_adjacency():
```
Comment (tomvanmele) [inline tests/compas/topology/test_unify_cycles.py]: why is this a face adjacency test and not also a test of unify cycles?

**Review by tomvanmele — COMMENTED** [review]
everything looks good but the tests are a bit confusing.
should we not include an assertion of some sorts?
also, if you know when the tests are supposed to fail, should we then not add a `@pytest.mark.xfail`?

**Comment by tomvanmele** [comment]
@romanarust perhaps just a few things to sort out regarding the test(s)

**Comment by romanarust** [comment]
> @romanarust perhaps just a few things to sort out regarding the test(s)

Thanks! I've added now some `assert` to the tests.

**Review by tomvanmele — APPROVED** [review]
LGTM
